FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1494 @@
+ */
+typedef struct subscriptionElement
+{
+    char pTopicFilter[ DEMO_BUFFER_SIZE ];
+    uint16_t topicFilterLength;
+    QueueHandle_t pResponseQueue;
+} SubscriptionElement_t;
+
+/**
+ * @brief An element for a task's response queue for received publishes.
+ *
+ * @note Since elements are copied to queues, this struct needs to hold
+ * buffers for the payload and topic of incoming publishes, as the original
+ * pointers are out of scope. When processing a publish from this struct,
+ * the `pTopicName` and `pPayload` pointers need to be set to point to the
+ * static buffers in this struct.
+ */
+typedef struct publishElement
+{
+    MQTTPublishInfo_t publishInfo;
+    uint8_t pPayload[ DEMO_BUFFER_SIZE ];
+    uint8_t pTopicName[ DEMO_BUFFER_SIZE ];
+} PublishElement_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Sends an MQTT Connect packet over the already connected TCP socket.
+ *
+ * @param[in] pxMQTTContext MQTT context pointer.
+ * @param[in] xNetworkContext Network context.
+ * @param[in] xCleanSession If a clean session should be established.
+ */
+static MQTTStatus_t prvMQTTConnect( MQTTContext_t * pxMQTTContext,
+                                    NetworkContext_t * pxNetworkContext,
+                                    bool xCleanSession );
+
+/**
+ * @brief Initialize context for a command.
+ *
+ * @param[in] pxContext Context to initialize.
+ */
+static void prvInitializeCommandContext( CommandContext_t * pxContext );
+
+/**
+ * @brief Add an operation to the list of pending acks.
+ *
+ * @param[in] usPacketId Packet ID of pending ack.
+ * @param[in] pCommand Copy of command that is expecting an ack.
+ */
+static void prvAddAck( uint16_t usPacketId,
+                       Command_t * pCommand );
+
+/**
+ * @brief Remove an operation from the list of pending acks and return it.
+ *
+ * @param[in] usPacketId Packet ID of incoming ack.
+ * @param[in] xRemove Flag indicating if an ack should be removed
+ *
+ * @return Stored information about the operation awaiting the ack.
+ */
+static AckInfo_t prvGetAck( uint16_t usPacketId,
+                            bool xRemove );
+
+/**
+ * @brief Add a subscription to the subscription list.
+ *
+ * @note Multiple tasks can be subscribed to the same topic. However, a single
+ * task may only subscribe to the same topic filter once.
+ *
+ * @param[in] pTopicFilter Topic filter of subscription.
+ * @param[in] topicFilterLength Length of topic filter.
+ * @param[in] pxQueue Response queue in which to enqueue received publishes.
+ */
+static void prvAddSubscription( const char * pTopicFilter,
+                                uint16_t topicFilterLength,
+                                QueueHandle_t pxQueue );
+
+/**
+ * @brief Remove a subscription from the subscription list.
+ *
+ * @note If the topic filter exists multiple times in the subscription list,
+ * then every instance of the subscription will be removed.
```

Comment:
How will that handle the case in which multiple tasks are subscribed to the same topic, but only one needs to be removed?

Reply:
It won't. That's an additional feature someone could add if they wanted, but it's outside the scope of this demo

Reply:
I understand the reason for doing it this way. Thanks